ada_feeding_perception/ada_feeding_perception/face_detection.py

```diff
@@ -0,0 +1,388 @@
+                        self.img_timestamp.nanosec / (10**9)
+                    )
+                    difference_array = np.append(
+                        difference_array, abs(depth_time - img_time)
+                    )
+                closest_depth = self.bridge.imgmsg_to_cv2(
+                    self.depth_buffer[difference_array.argmin()],
+                    desired_encoding="passthrough",
+                )
+
+                # Retrieve the depth value averaged over all mouth coordinates
+                depth_sum = 0
+                for point in self.img_mouth_points:
+                    depth_sum += closest_depth[int(point[1])][int(point[0])]
+                depth = depth_sum / float(len(self.img_mouth_points))
+
+                # Create target 3d point, with mm measurements converted to m
+                mouth_location = PointStamped()
+                mouth_location.header = msg.header
+                mouth_location.point.x = (
+                    float(depth) * (float(u) - self.camera_matrix[2])
+                ) / (1000.0 * self.camera_matrix[0])
+                mouth_location.point.y = (
+                    float(depth) * (float(v) - self.camera_matrix[5])
+                ) / (1000.0 * self.camera_matrix[4])
+                mouth_location.point.z = float(depth) / 1000.0
+
+                face_detection_msg.detected_mouth_center = mouth_location
+
+            # Publish 3d point
+            face_detection_msg.is_face_detected = self.is_face_detected
+            self.publisher_results.publish(face_detection_msg)
+
+    def camera_callback(self, msg: CompressedImage):
+        """
+        Callback function for the camera feed. If face detection is on, this
+        function will detect faces in the image and store the location of
+        the largest detected face/mouth.
+        """
+        # pylint: disable=too-many-locals
+        # Two variables over the limit is fine, needed to find both faces and mouths
+        image_rgb = cv2.imdecode(np.frombuffer(msg.data, np.uint8), cv2.IMREAD_COLOR)
+        with self.is_on_lock:
+            is_on = self.is_on
+
+        if is_on:
+            self.is_face_detected = False
+
+            # Convert image to Grayscale
+            image_gray = cv2.cvtColor(image_rgb, cv2.COLOR_RGB2GRAY)
+
+            # Create an instance of the Face Detection Cascade Classifier
+            detector = cv2.CascadeClassifier(self.face_model_path)
+            # Detect faces using the haarcascade classifier on the grayscale image
+            faces = detector.detectMultiScale(image_gray)
+
+            # Check if a face has been detected
+            if len(faces) > 0:
+                self.is_face_detected = True
+
+            if self.is_face_detected:
+                # Create an instance of the Facial landmark Detector with the model
+                landmark_detector = cv2.face.createFacemarkLBF()
+                landmark_detector.loadModel(self.landmark_model_path)
```

Comment:
Also wondering if this can be initialized/loaded just once in `__init__` instead of every time in the callback. Loading the models probably slows the callback down quite a bit.